CloudFront Signer: load private key once

danielbeardsley opened the pull request:
openss_sign() is abysmally slow. On a fast CPU with a small amount of
data it takes 1.5ms of CPU time.

Turns out, half of that was parsing the private key. If we have
openssl parse the key ahead of time, this operation takes half the
time.

This change brought it from 1.5ms to 0.7ms per call, still really
slow but much better.

Note: We started running into performance issues with *lots* of image
urls being rendered on long pages.

danielbeardsley committed:
CloudFront Signer: load private key once

openss_sign() is abysmally slow. On a fast CPU with a small amount of
data it takes 1.5ms of CPU time.

Turns out, half of that was parsing the private key. If we have
openssl parse the key ahead of time, this operation takes half the
time.

This change brought it from 1.5ms to 0.7ms per call, still really
slow but much better.

Note: We started running into performance issues with *lots* of image
urls being rendered on long pages.

danielbeardsley commented:
This is the local version of aws/aws-sdk-php#1752 FYI

danielbeardsley commented:
QA 👍  (it passed CI on the parent repo).

BaseInfinity commented:
CR 👙

davidrans commented:
CR ☕️

addison-grant commented:
Are you sure that tests on the parent repo's CI actually test this code? Coverage there isn't 100% and I don't see where it says that this code is covered. In fact, looking at the code cov link, it looks like that code isn't covered... Not sure if I am interpreting it correctly.

Anyways, I checked out this branch in our vendor/ dir with https://github.com/iFixit/ifixit/pull/27951 checked out (ran `composer dumpautoload`; not sure if this was necessary to pick up code changes), and ran through the steps there to make CdnUrls work on cominor, and it looks to me like it works. I haven't confirmed that I am actually using the code yet.

addison-grant commented:
```diff
>> git diff
diff --git a/src/CloudFront/Signer.php b/src/CloudFront/Signer.php
index d8e14be5..5a9f9a97 100644
--- a/src/CloudFront/Signer.php
+++ b/src/CloudFront/Signer.php
@@ -21,6 +21,7 @@ class Signer
      */
     public function __construct($keyPairId, $privateKey)
     {
+       \Log::debug("In Danny's code");
         if (!extension_loaded('openssl')) {
             //@codeCoverageIgnoreStart
             throw new \RuntimeException('The openssl extension is required to '
```

That log got executed, so I believe Danny's code got run, and I think it's working.

danielbeardsley committed:
SignerTest: add test for bad private key

This is mostly to ensure that key parsing is happening
during construction instead of on every call to ->sign()

We can't *really* enforce that the key isn't parsed on every call
without exposing more of the implementation, so we'll settle on this.

danielbeardsley committed:
CloudFront Signer Optimization: add changelog entry

danielbeardsley committed:
Cloudfront Signer Test: clean up afterward
